// @graphql-toolkit/url-loader/dist/esnext/load-from-url.js
import { introspectionQuery, buildClientSchema, parse } from 'graphql';
import { printSchemaWithDirectives } from '@graphql-toolkit/common';
import { isUri } from 'valid-url';
export class UrlLoader {
    loaderId() {
        return 'url';
    }
    async canLoad(pointer, options) {
        return !!isUri(pointer);
    }
    async load(pointer, options) {
        let headers = {};
        let fetch;
        if (options) {
            if (Array.isArray(options.headers)) {
                headers = options.headers.reduce((prev, v) => ({ ...prev, ...v }), {});
            }
            else if (typeof options.headers === 'object') {
                headers = options.headers;
            }
            if (options.fetch) {
                fetch = options.fetch;
            }
        }
        if (!fetch) {
            fetch = (await import('cross-fetch')).fetch;
        }
        let extraHeaders = {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            ...headers,
        };
        const response = await fetch(pointer, {
            method: 'POST',
            body: JSON.stringify({
                query: introspectionQuery,
            }),
            headers: extraHeaders,
        });
        const body = await response.json();
        let errorMessage;
        if (body.errors && body.errors.length > 0) {
            errorMessage = body.errors.map((item) => item.message).join(', ');
        }
        else if (!body.data) {
            errorMessage = body;
        }
        if (errorMessage) {
            throw 'Unable to download schema from remote: ' + errorMessage;
        }
        if (!body.data.__schema) {
            throw new Error('Invalid schema provided!');
        }
        const asSchema = buildClientSchema(body.data);
        const printed = printSchemaWithDirectives(asSchema);
        return {
            location: pointer,
            document: parse(printed),
        };
    }
}
