// @graphql-toolkit/graphql-file-loader/dist/commonjs/load-from-gql-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const path_1 = require("path");
const GQL_EXTENSIONS = ['.gql', '.graphql', '.graphqls'];
class GraphQLFileLoader {
    loaderId() {
        return 'graphql-file';
    }
    async canLoad(pointer, options) {
        const extension = path_1.extname(pointer).toLowerCase();
        return GQL_EXTENSIONS.includes(extension);
    }
    async load(pointer, options) {
        const { readFileSync } = eval(`require('fs')`);
        const normalizedFilePath = path_1.isAbsolute(pointer) ? pointer : path_1.resolve(options.cwd || process.cwd(), pointer);
        const content = readFileSync(normalizedFilePath, 'utf-8').trim();
        if (content && content !== '') {
            if (!options.skipGraphQLImport && /^\#.*import /i.test(content.trimLeft())) {
                const { importSchema } = eval(`require('graphql-import')`);
                const importedSchema = importSchema(normalizedFilePath);
                return {
                    location: pointer,
                    document: graphql_1.parse(importedSchema),
                };
            }
            else {
                return {
                    location: pointer,
                    document: graphql_1.parse(new graphql_1.Source(content, pointer)),
                };
            }
        }
        return null;
    }
}
exports.GraphQLFileLoader = GraphQLFileLoader;
