// @graphql-toolkit/url-loader/commonjs/load-from-url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const common_1 = require("@graphql-toolkit/common");
const valid_url_1 = require("valid-url");
const cross_fetch_1 = require("cross-fetch");
class UrlLoader {
    loaderId() {
        return 'url';
    }
    async canLoad(pointer, options) {
        return !!valid_url_1.isUri(pointer);
    }
    async load(pointer, options) {
        let headers = {};
        let fetch = cross_fetch_1.fetch;
        let method = 'POST';
        if (options) {
            if (Array.isArray(options.headers)) {
                headers = options.headers.reduce((prev, v) => ({ ...prev, ...v }), {});
            }
            else if (typeof options.headers === 'object') {
                headers = options.headers;
            }
            if (options.customFetch) {
                if (typeof options.customFetch === 'string') {
                    const [moduleName, fetchFnName] = options.customFetch.split('#');
                    fetch = await Promise.resolve().then(() => require(moduleName)).then(module => (fetchFnName ? module[fetchFnName] : module));
                }
            }
            if (options.method) {
                method = options.method;
            }
        }
        let extraHeaders = {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            ...headers,
        };
        const response = await fetch(pointer, {
            method,
            ...(method === 'POST'
                ? {
                    body: JSON.stringify({
                        query: graphql_1.introspectionQuery,
                    }),
                }
                : {}),
            headers: extraHeaders,
        });
        const body = await response.json();
        let errorMessage;
        if (body.errors && body.errors.length > 0) {
            errorMessage = body.errors.map((item) => item.message).join(', ');
        }
        else if (!body.data) {
            errorMessage = body;
        }
        if (errorMessage) {
            throw new Error('Unable to download schema from remote: ' + errorMessage);
        }
        if (!body.data.__schema) {
            throw new Error('Invalid schema provided!');
        }
        const asSchema = graphql_1.buildClientSchema(body.data);
        const printed = common_1.printSchemaWithDirectives(asSchema);
        return {
            location: pointer,
            document: graphql_1.parse(printed),
        };
    }
}
exports.UrlLoader = UrlLoader;
