// @graphql-toolkit/code-file-loader/esnext/load-from-code-file.js
import { GraphQLSchema, parse, buildClientSchema, Source as GraphQLSource } from 'graphql';
import { resolve, isAbsolute, extname } from 'path';
import { extractDocumentStringFromCodeFile } from './extract-document-string-from-code-file';
import { debugLog, printSchemaWithDirectives, asArray, fixWindowsPath } from '@graphql-toolkit/common';
import { readFileSync, existsSync } from 'fs';
import * as isValidPath from 'is-valid-path';
function isSchemaText(obj) {
    return typeof obj === 'string';
}
function isWrappedSchemaJson(obj) {
    const json = obj;
    return json.data !== undefined && json.data.__schema !== undefined;
}
function isSchemaJson(obj) {
    const json = obj;
    return json !== undefined && json.__schema !== undefined;
}
function isSchemaObject(obj) {
    return obj instanceof GraphQLSchema;
}
function isSchemaAst(obj) {
    return obj.kind !== undefined;
}
function resolveExport(fileExport) {
    if (isSchemaObject(fileExport)) {
        return parse(printSchemaWithDirectives(fileExport));
    }
    else if (isSchemaText(fileExport)) {
        return parse(fileExport);
    }
    else if (isWrappedSchemaJson(fileExport)) {
        const asSchema = buildClientSchema(fileExport.data);
        const printed = printSchemaWithDirectives(asSchema);
        return parse(printed);
    }
    else if (isSchemaJson(fileExport)) {
        const asSchema = buildClientSchema(fileExport);
        const printed = printSchemaWithDirectives(asSchema);
        return parse(printed);
    }
    else if (isSchemaAst(fileExport)) {
        return fileExport;
    }
    return null;
}
async function tryToLoadFromExport(rawFilePath) {
    let filePath = rawFilePath;
    try {
        filePath = fixWindowsPath(filePath);
        if (require && require.cache) {
            filePath = require.resolve(filePath);
            if (require.cache[filePath]) {
                delete require.cache[filePath];
            }
        }
        const rawExports = await import(filePath);
        if (rawExports) {
            let rawExport = rawExports.default || rawExports.schema || rawExports;
            if (rawExport) {
                let exportValue = await rawExport;
                exportValue = await (exportValue.default || exportValue.schema || exportValue.typeDefs || exportValue);
                try {
                    return resolveExport(exportValue);
                }
                catch (e) {
                    throw new Error('Exported schema must be of type GraphQLSchema, text, AST, or introspection JSON.');
                }
            }
            else {
                throw new Error(`Invalid export from export file ${filePath}: missing default export or 'schema' export!`);
            }
        }
        else {
            throw new Error(`Invalid export from export file ${filePath}: empty export!`);
        }
    }
    catch (e) {
        throw new Error(`Unable to load from file "${filePath}": ${e.message}`);
    }
}
async function tryToLoadFromCodeAst(filePath, options) {
    const content = readFileSync(filePath, 'utf-8');
    const foundDoc = await extractDocumentStringFromCodeFile(new GraphQLSource(content, filePath), options || {});
    if (foundDoc) {
        return parse(foundDoc);
    }
    else {
        return null;
    }
}
const CODE_FILE_EXTENSIONS = ['.ts', '.tsx', '.js', '.jsx'];
export class CodeFileLoader {
    loaderId() {
        return 'code-file';
    }
    async canLoad(pointer, options) {
        if (isValidPath(pointer)) {
            const extension = extname(pointer).toLowerCase();
            if (CODE_FILE_EXTENSIONS.includes(extension)) {
                const normalizedFilePath = isAbsolute(pointer) ? pointer : resolve(options.cwd || process.cwd(), pointer);
                if (existsSync(normalizedFilePath)) {
                    return true;
                }
            }
        }
        return false;
    }
    async load(pointer, options) {
        let loaded = null;
        const normalizedFilePath = isAbsolute(pointer) ? pointer : resolve(options.cwd || process.cwd(), pointer);
        try {
            const result = await tryToLoadFromCodeAst(normalizedFilePath, options);
            if (result) {
                loaded = result;
            }
        }
        catch (e) {
            debugLog(`Failed to load schema from code file "${normalizedFilePath}" using AST: ${e.message}`);
            throw e;
        }
        if (!loaded && !options.noRequire) {
            if (options && options.require) {
                await Promise.all(asArray(options.require).map(m => import(m)));
            }
            loaded = await tryToLoadFromExport(normalizedFilePath);
        }
        return {
            document: loaded,
            location: normalizedFilePath,
        };
    }
}
